lib/c-api/src/wasm_c_api/engine/config/sys.rs

```diff
@@ -0,0 +1,436 @@
+) {
+    // XXX: Debatable.
+    //
+    // On one hand, if a user calls this function, it means that they want to use the `sys` engine.
+    // On the other, this can make the rest of the code fail if the user calls functions that use
+    // values generated from other engines before creating a new one.
+    //
+    // The other option would be to check if the current engine is indeed `sys`, and do nothing if
+    // it is not. We can't, however, return an explicit error from this function, so the failure
+    // would be silent.
+    config.engine = wasmer_engine_t::UNIVERSAL;
+
+    if let wasmer_engine_config_t::Sys(ref mut c) = config.engine_config {
+        c.middlewares.push(*middleware);
+    } else {
+        let sys_config = wasmer_sys_engine_config_t {
+            middlewares: vec![*middleware],
+            ..Default::default()
+        };
+
+        config.engine_config = wasmer_engine_config_t::Sys(sys_config);
+    }
+}
+
+/// Updates the configuration to enable NaN canonicalization.
+///
+/// This is a Wasmer-specific function.
+///
+/// # Example
+///
+/// ```rust
+/// # use wasmer_inline_c::assert_c;
+/// # fn main() {
+/// #    (assert_c! {
+/// # #include "tests/wasmer.h"
+/// #
+/// int main() {
+///     // Create the configuration.
+///     wasm_config_t* config = wasm_config_new();
+///
+///     // Enable NaN canonicalization.
+///     wasm_config_sys_canonicalize_nans(config, true);
+///
+///     // Create the engine.
+///     wasm_engine_t* engine = wasm_engine_new_with_config(config);
+///
+///     // Check we have an engine!
+///     assert(engine);
+///
+///     // Free everything.
+///     wasm_engine_delete(engine);
+///
+///     return 0;
+/// }
+/// #    })
+/// #    .success();
+/// # }
+/// ```
+#[no_mangle]
+pub extern "C" fn wasm_config_sys_canonicalize_nans(config: &mut wasm_config_t, enable: bool) {
+    // XXX: Debatable.
+    //
+    // On one hand, if a user calls this function, it means that they want to use the `sys` engine.
+    // On the other, this can make the rest of the code fail if the user calls functions that use
+    // values generated from other engines before creating a new one.
+    //
+    // The other option would be to check if the current engine is indeed `sys`, and do nothing if
+    // it is not. We can't, however, return an explicit error from this function, so the failure
+    // would be silent.
+    config.engine = wasmer_engine_t::UNIVERSAL;
+
+    if let wasmer_engine_config_t::Sys(ref mut c) = config.engine_config {
+        c.nan_canonicalization = enable;
+    } else {
+        let sys_config = wasmer_sys_engine_config_t {
+            nan_canonicalization: enable,
+            ..Default::default()
+        };
+        config.engine_config = wasmer_engine_config_t::Sys(sys_config);
+    }
+}
+
+/// Create a new  [`wasm_engine_t`] backed by a `sys` engine.
+pub fn wasm_sys_engine_new_with_config(config: wasm_config_t) -> Option<Box<wasm_engine_t>> {
```

Comment:
This should not be public

Reply:
[5c08932](https://github.com/wasmerio/wasmer/pull/5443/commits/5c08932b8602ade591afaabe379c52da6f8d57d4)